src/modules/sync/diff/contentType.ts

```diff
@@ -0,0 +1,87 @@
+import { ContentTypeSyncModel } from "../types/fileContentModel.js";
+import { SyncGuidelinesElement } from "../types/syncModel.js";
+import {
+  baseHandler,
+  Handler,
+  makeArrayHandler,
+  makeLeafObjectHandler,
+  makeObjectHandler,
+  makeUnionHandler,
+  optionalHandler,
+} from "./combinators.js";
+import {
+  makeAssetElementHandler,
+  makeCustomElementHandler,
+  makeDateTimeElementHandler,
+  makeGuidelinesElementHandler,
+  makeLinkedItemsElementHandler,
+  makeMultiChoiceElementHandler,
+  makeNumberElementHandler,
+  makeRichTextElementHandler,
+  makeSnippetElementHandler,
+  makeSubpagesElementHandler,
+  makeTaxonomyElementHandler,
+  makeTextElementHandler,
+  makeUrlSlugElementHandler,
+} from "./modelElements.js";
+import { transformGuidelinesElementToAddModel } from "./transformToAddModel.js";
+
+type HandleContentTypeParams = Readonly<{
+  targetItemsByCodenames: ReadonlyMap<string, Readonly<{ id: string; codename: string }>>;
+  targetAssetsByCodenames: ReadonlyMap<string, Readonly<{ id: string; codename: string }>>;
+}>;
+
+export const makeContentTypeHandler = (
+  params: HandleContentTypeParams,
+): Handler<ContentTypeSyncModel> =>
+  makeObjectHandler({
+    name: baseHandler,
+    content_groups: optionalHandler(makeArrayHandler(
+      g => g.codename,
+      makeObjectHandler({
+        name: baseHandler,
+      }),
+    )),
+    elements: {
+      contextfulHandler: ({ source, target }) => {
+        const ctx = {
+          ...params,
+          targetAssetCodenames: new Set(params.targetAssetsByCodenames.keys()),
+          targetItemCodenames: new Set(params.targetItemsByCodenames.keys()),
+          sourceTypeOrSnippet: source,
+          targetTypeOrSnippet: target,
+        };
+
+        return makeArrayHandler(
+          el => el.codename,
+          makeUnionHandler("type", {
+            number: makeNumberElementHandler(ctx),
+            text: makeTextElementHandler(ctx),
+            asset: makeAssetElementHandler(ctx),
+            guidelines: makeGuidelinesElementHandler(ctx),
+            custom: makeCustomElementHandler(ctx),
+            snippet: makeSnippetElementHandler(ctx),
+            subpages: makeSubpagesElementHandler(ctx),
+            taxonomy: makeTaxonomyElementHandler(ctx),
+            url_slug: makeUrlSlugElementHandler(ctx),
+            date_time: makeDateTimeElementHandler(ctx),
+            rich_text: makeRichTextElementHandler(ctx),
+            modular_content: makeLinkedItemsElementHandler(ctx),
+            multiple_choice: makeMultiChoiceElementHandler(ctx),
+          }),
```

Comment:
nitpick: I would extract this functions

Reply:
I would leave it, if you agree, as is since the handler is not too long and extracting the inner handler outside would make it loose the context of the parent handler.